### PR TITLE
chore: add type definitions for the `eslint-config-eslint` package

### LINKED ISSUE #19043: Change Request: Add type definitions for `eslint-config-eslint`
### ESLint version

v9

### What problem do you want to solve?

Add type definitions for `eslint-config-eslint`.

### What do you think is the correct solution?

Manually add type definitions for `eslint-config-eslint`.

### Participation

- [x] I am willing to submit a pull request for this change.

### Additional comments

_No response_

### PR DESCRIPTION
**Pull request by aryaemami59**
#### Prerequisites checklist

- [X] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[X] Other, please explain:

## **This PR**:

- [X] Adds type definitions for the `eslint-config-eslint` package.
- [X] Resolves #19043.

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -116,12 +116,24 @@ jobs:
   test_types:
     name: Types Test
     runs-on: ubuntu-latest
+
+    strategy:
+      matrix:
+        package:
+          [{ name: eslint-config-eslint, directory: packages/eslint-config-eslint }]
```
Comment (nzakas) [inline .github/workflows/ci.yml:123]: I might be missing something, but with this change, won't we end up skipping the types test in the project root? Don't we need another entry in the array?

File: eslint.config.js [hunk 2]
```diff
@@ -70,6 +70,9 @@ function createInternalFilesPatterns(pattern = null) {
         }));
 }
 
+/**
+ * @type {import("eslint").Linter.Config[]}
```
Comment (nzakas) [inline eslint.config.js:74]: Should import from `lib/types` instead?

File: packages/eslint-config-eslint/formatting.js [hunk 3]
```diff
@@ -1,123 +1,135 @@
 "use strict";
 
+const rules = {
```
Comment (nzakas) [inline packages/eslint-config-eslint/formatting.js:3]: What's the purpose of this change?

**Review by nzakas — COMMENTED** [review]
Overall LGTM. Just a few questions.

File: eslint.config.js (hunk 2 above)
Comment (aryaemami59) [reply eslint.config.js:74]: It resolves to `lib/types/index` so they're the same thing but I'll change it to be more explicit.

File: packages/eslint-config-eslint/formatting.js (hunk 3 above)
Comment (aryaemami59) [reply packages/eslint-config-eslint/formatting.js:3]: It's so that we can use named imports in ESM files:

```ts
import { rules } from "eslint-config-eslint/formatting";

export default [...eslintConfigESLint, rules] satisfies Linter.Config[];
```

File: .github/workflows/ci.yml (hunk 1 above)
Comment (aryaemami59) [reply .github/workflows/ci.yml]: You're right, I fixed it.

File: packages/eslint-config-eslint/nodejs.js [hunk 4]
```diff
@@ -1,14 +1,24 @@
 "use strict";
 
-const recommendedScriptConfig = require("eslint-plugin-n/configs/recommended-script");
-const recommendedModuleConfig = require("eslint-plugin-n/configs/recommended-module");
+const {
+    configs: {
+        "flat/recommended-script": recommendedScriptConfig,
+        "flat/recommended-module": recommendedModuleConfig
+    }
+} = require("eslint-plugin-n");
```
Comment (aryaemami59) [inline packages/eslint-config-eslint/nodejs.js:8]: [**"eslint-plugin-n/configs/recommended-script"**](https://github.com/eslint-community/eslint-plugin-n/blob/master/configs/recommended-script.js#L4) and [**"eslint-plugin-n/configs/recommended-module"**](https://github.com/eslint-community/eslint-plugin-n/blob/master/configs/recommended-script.js#L4) are deprecated.

File: packages/eslint-config-eslint/package.json [hunk 5]
```diff
@@ -21,7 +50,8 @@
     "cjs.js",
     "formatting.js",
     "index.js",
-    "nodejs.js"
+    "nodejs.js",
+    "types"
```
Comment (snitin315) [inline packages/eslint-config-eslint/package.json:54]: We don't need to include `ambient-module-declarations`?

**Review by snitin315 — COMMENTED** [review]
Overall looks good. Few comments. Can you look into the failing CI as well?

File: packages/eslint-config-eslint/ambient-module-declarations/@eslint-community-eslint-plugin-eslint-comments.d.ts [hunk 6]
```diff
@@ -0,0 +1,23 @@
+declare module "@eslint-community/eslint-plugin-eslint-comments" {
```
[the remark this answers is not in the document]
Comment (aryaemami59) [reply packages/eslint-config-eslint/ambient-module-declarations/@eslint-community-eslint-plugin-eslint-comments.d.ts:1]: No because We don't want them to be included in the final package tarball.

File: packages/eslint-config-eslint/ambient-module-declarations/eslint-plugin-jsdoc.d.ts [hunk 7]
```diff
@@ -0,0 +1,23 @@
+declare module "eslint-plugin-jsdoc" {
+  import type { ESLint, Linter } from "eslint";
+
+  const eslintPluginJsdoc: ESLint.Plugin & {
+      configs: Record<
+          | "recommended"
```
[the remark this answers is not in the document]
Comment (aryaemami59) [reply packages/eslint-config-eslint/ambient-module-declarations/eslint-plugin-jsdoc.d.ts:6]: No we don't have to worry about these as they are just added for intellisense for us, but I will open issues in `eslint-plugin-jsdoc` and `eslint-plugin-unicorn` regardless.

File: packages/eslint-config-eslint/package.json (hunk 5 above)
Comment (aryaemami59) [reply packages/eslint-config-eslint/package.json:54]: No. The `ambient-module-declarations` is just for us.

File: packages/eslint-config-eslint/formatting.js (hunk 3 above)
Comment (aryaemami59) [reply packages/eslint-config-eslint/formatting.js:6]: In an ESM file we will be able to do:

```ts
import { rules } from "eslint-config-eslint/formatting";

export default [...eslintConfigESLint, rules] satisfies Linter.Config[];
```

Without the provided changes, this would fail.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -57,6 +57,8 @@ jobs:
         uses: trunk-io/trunk-action@v1
         with:
 
+          save-annotations: false
+
```
Comment (aryaemami59) [inline .github/workflows/ci.yml:61]: Not sure why but I had to disable `save-annotations` to make CI happy.

Comment (snitin315) [reply .github/workflows/ci.yml:61]: What was the error with  `save-annotations` enabled?

Comment (aryaemami59) [reply .github/workflows/ci.yml:61]: https://github.com/eslint/eslint/actions/runs/11533071521/job/32105566286

Comment (nzakas) [reply .github/workflows/ci.yml:61]: That looks like an error in Trunk that we should get checked out.

@det

Comment (fasttime) [reply .github/workflows/ci.yml:61]: This is the same problem we had in the main branch before pinning `trunk-action` to v1.1.17: https://github.com/trunk-io/trunk-action/pull/266.

**Review by JoshuaKGoldberg — APPROVED** [review]
LGTM! Just one suggestion about redundant properties, but nothing blocking IMO.

File: packages/eslint-config-eslint/tsconfig.json [hunk 9]
```diff
@@ -0,0 +1,20 @@
+{
+    "compilerOptions": {
+        "allowSyntheticDefaultImports": false,
+        "esModuleInterop": false,
```
[the remark this answers is not in the document]
Comment (aryaemami59) [reply packages/eslint-config-eslint/tsconfig.json:4]: Actually [**`esModuleInterop`**](https://www.typescriptlang.org/tsconfig/#esModuleInterop) is automatically set to `true` if [**`module`**](https://www.typescriptlang.org/tsconfig/#module) is set to **`Node16`** or **`NodeNext`**. But this makes me wonder if explicitly disable [**`esModuleInterop`**](https://www.typescriptlang.org/tsconfig/#esModuleInterop), would it still techinically be set to `true` since we have [**`module`**](https://www.typescriptlang.org/tsconfig/#module) set to **`NodeNext`**?

[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/eslint-config-eslint/tsconfig.json:4]: Since it's just being used for type-checking and not transpilation, I'd say as long as there are no type errors it's 👍

[the remark this answers is not in the document]
Comment (aryaemami59) [reply packages/eslint-config-eslint/tsconfig.json:4]: Awesome, so you still think I should remove them?

[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/eslint-config-eslint/tsconfig.json:4]: I do, yup

[the remark this answers is not in the document]
Comment (aryaemami59) [reply packages/eslint-config-eslint/tsconfig.json:4]: I tried it, removing them makes the type tests fail.

[the remark this answers is not in the document]
Comment (JoshuaKGoldberg) [reply packages/eslint-config-eslint/tsconfig.json:4]: In that case, I suppose I'm wrong to suggest removing it 😄

File: packages/eslint-config-eslint/formatting.js (hunk 3 above)
Comment (aladdin-add) [reply packages/eslint-config-eslint/formatting.js:6]: is it supported and documented somewhere? As far as I can remember, cjs should be treated as a default export.

Comment (aladdin-add) [reply packages/eslint-config-eslint/formatting.js:6]: it's very helpful, thanks for the explanation!

I agreed with @mdjermanovic said in https://github.com/eslint/eslint/issues/19110#issuecomment-2460723401.
it's explicitly stated the default export usage, and it's the officially recommended; we do not guarantee that non-recommended usage will work.

**Review by aladdin-add — APPROVED** [review]
LGTM. Would like another review before merging.

File: packages/eslint-config-eslint/package.json [hunk 10]
```diff
@@ -36,6 +66,20 @@
     "eslint-plugin-n": "^17.11.1",
     "eslint-plugin-unicorn": "^52.0.0"
   },
+  "devDependencies": {
+    "@arethetypeswrong/cli": "^0.16.4",
+    "eslint": "^9.14.0",
+    "eslint-config-eslint": "file:",
```
Comment (fasttime) [inline packages/eslint-config-eslint/package.json:72]: As explained in https://nodejs.org/api/packages.html#self-referencing-a-package-using-its-name, this dependency is unnecessary: `eslint-config-eslint` will be automatically resolved to the current package because it matches the `"name"` property in `package.json` and an `"exports"` property is present.

File: packages/eslint-config-eslint/ambient-module-declarations/@eslint-community-eslint-plugin-eslint-comments.d.ts
Comment (fasttime) [inline packages/eslint-config-eslint/ambient-module-declarations/@eslint-community-eslint-plugin-eslint-comments.d.ts:1]: What is the purpose of these ambient declarations? VSCode seems to ignore them completely. If I delete the `ambient-module-declarations` directory I can see the types inferred correctly in the IDE and all scripts in `package.json` are still passing:

image

**Review by fasttime — APPROVED** [review]
LGTM, thanks! Since this is a `feat:` pull request we'll have to wait until #19096 is closed before merging.

File: packages/eslint-config-eslint/formatting.js
Comment (mdjermanovic) [inline packages/eslint-config-eslint/formatting.js:1]: If the changes in this file are made only to allow for `import { rules } from "eslint-config-eslint/formatting";`, then I think we should revert them, because `import { rules } from "eslint-config-eslint/formatting";` is not supposed to work. This module exports a config object with a `rules` property, not named export `rules`.

File: packages/eslint-config-eslint/types/formatting.d.ts [hunk 11]
```diff
@@ -0,0 +1,3 @@
+import type { Linter } from "eslint";
+
+export declare const rules: NonNullable<Linter.Config["rules"]>;
```
Comment (mdjermanovic) [inline packages/eslint-config-eslint/types/formatting.d.ts:3]: I think this doesn't reflect the intended usage, which is `import eslintConfigESLintFormatting from "eslint-config-eslint/formatting";`:

https://github.com/eslint/eslint/tree/main/packages/eslint-config-eslint#formatting-config

 This module exports a `Linter.Config` object, not named export `rules`.

File: packages/eslint-config-eslint/tests/types/types.test-d.mts [hunk 12]
```diff
@@ -0,0 +1,20 @@
+import type { Linter } from "eslint";
+
+import * as eslintConfigESLint from "eslint-config-eslint";
+import * as eslintConfigESLintBase from "eslint-config-eslint/base";
+import * as eslintConfigESLintCJS from "eslint-config-eslint/cjs";
+import * as eslintConfigESLintFormatting from "eslint-config-eslint/formatting";
```
[the remark this answers is not in the document]
Comment (fasttime) [reply packages/eslint-config-eslint/tests/types/types.test-d.mts:6]: > I'm not sure how these tests are passing because, for example, `eslintConfigESLint` evaluates to `{ default: [...] }`.

I think that's because the [`allowSyntheticDefaultImports`](https://www.typescriptlang.org/tsconfig/#allowSyntheticDefaultImports) is set to `false` in the TypeScript config. We could change that option to `true` to disallow the `import *` syntax in the types tests if it's confusing.

**Comment by mdjermanovic** [comment]
I changed the tag to `chore` because all commits in this repo appear in changelogs for the `eslint` package but this isn't a new feature in `eslint`.

Regardless, when releasing `eslint-config-eslint` with this change, I think we should choose __major__ as RELEASE_TYPE, to bump the version to 12.0.0, because adding `peerDependencies` is a breaking change.

**Comment by fasttime** [comment]
If we are going to release this change in a new major version I think we could as well set the `eslint` version in `peerDependencies` to `"^9.10.0"`, since v9.10.0 is the first ESLint version that includes types.

File: packages/eslint-config-eslint/tests/types/types.test-d.mts (hunk 12 above)
[the remark this answers is not in the document]
Comment (mdjermanovic) [reply packages/eslint-config-eslint/tests/types/types.test-d.mts:6]: > I think that's because the [`allowSyntheticDefaultImports`](https://www.typescriptlang.org/tsconfig/#allowSyntheticDefaultImports) is set to `false` in the TypeScript config. We could change that option to `true` to disallow the `import *` syntax in the types tests if it's confusing.

Sounds good to me 👍

**Comment by nzakas** [comment]
Is this ready for another review?

**Comment by fasttime** [comment]
It looks like https://github.com/eslint/eslint/pull/19050#discussion_r1845564637 hasn't been addressed yet.

**Comment by aryaemami59** [comment]
@fasttime Done.

@nzakas Yes.

**Review by mdjermanovic — DISMISSED** [review]
LGTM, thanks! Leaving open for @fasttime to re-review.

File: packages/eslint-config-eslint/tsconfig.json [hunk 13]
```diff
@@ -0,0 +1,19 @@
+{
+    "compilerOptions": {
+        "allowSyntheticDefaultImports": false,
```
Comment (fasttime) [inline packages/eslint-config-eslint/tsconfig.json:3]:
```suggestion
        "allowSyntheticDefaultImports": true,
```

I think the idea was to set `allowSyntheticDefaultImports` to `true` to enforce the syntax `import eslintConfigESLint from "eslint-config-eslint";` over `import * as eslintConfigESLint from "eslint-config-eslint";` in our tests.

`allowSyntheticDefaultImports` makes TypeScript behave more similarly to Node.js when it imports a CommonJS file, and that's what the team is more familiar with.

File: packages/eslint-config-eslint/tests/types/types.test-d.mts
Comment (fasttime) [inline packages/eslint-config-eslint/tests/types/types.test-d.mts:1]: Just asking: is this file named with `-d`  because it tests declaration files (`.d.ts`)? It took me a second look to realize that the file extension is just `.mts`.

File: packages/eslint-config-eslint/tests/types/types.test-d.ts
Comment (fasttime) [inline packages/eslint-config-eslint/tests/types/types.test-d.ts:1]: Same question as above about `-d` in the file name.

Comment (aryaemami59) [reply packages/eslint-config-eslint/tests/types/types.test-d.ts:1]: See [**this comment**](https://github.com/eslint/eslint/pull/19050#discussion_r1873812918)

**Review by fasttime — APPROVED** [review]
LGTM, thanks!
